WordPress/WordPressTest/ReaderReblogActionTests.swift

```diff
@@ -145,3 +141,14 @@ class ReblogFormatterTests: XCTestCase {
                        wpImage)
     }
 }
+
+private class MockViewController: UIViewController {
+
+    var presentExpectation: XCTestExpectation?
+
+    override func present(_ viewControllerToPresent: UIViewController, animated flag: Bool, completion: (() -> Void)? = nil) {
+        presentExpectation?.fulfill()
+        super.present(viewControllerToPresent, animated: flag, completion: completion)
+    }
+
+}
```

Comment:
I can see this being used in other tests, too. (I'm not actually sure there are other use cases, but I assume this mustn't be the only test calling `present...`.

How would you feel about making it "stricter"? I.e. require the expectation to be set, as that's the whole point why one would use this `UIViewController` subclass.

It could be done by `throw`ing if `presentExpectation` is `nil`. Or, we could provide a new, dedicated initializer, although, I always forget whether you can make the other ones `private` when working with `UIViewController` subclasses.

Another idea could be to move the whole expectation logic within the object, and make it expose a method to verify the presentation occurred. That would also be truer to the "mock" name, defined [in the xunit patterns](http://xunitpatterns.com/Mocks,%20Fakes,%20Stubs%20and%20Dummies.html) as a test double that verifies "indirect outputs of SUT" by verifying "correctness against expectations". (See my thoughts on test double naming [here](https://mokacoding.com/blog/swift-test-doubles/#what-about-mocks)).

```swift
let presenterMock = PresenterMockViewController()
...
presenter.presentReblog(blogService: blogService!, readerPost: readerPost!, origin: presenterMock)
...
presenterMock.expectPresentationToOccurr()

// In PresenterMockViewController
func expectPresentationToOccurr(file: ... = #file, line: ... = #line) {
  guard let presentExpectation = self.presentExpectation else { /* fail */ }

  waitForExpectations([presentExpectation], timeout: ...) { error in
    guard let error = error else { return }
    XCTFail("present was not called", file: file, line: line)
  }
}
```

Reply:
I agree with your thinking here—it should do what a "mock" type does if it's called a "mock". I tried to come up with some other ways to do this, like using OCMock(I couldn't get it to work in a Swift code), or capture the presented view controller and make assertions on it in the test function. But eventually I decided to go with what I had here, simply because that's the pattern used in the test class here (also a few other test classes that I ran into).

I do worry about having mixed patterns, especially if they are in the same file. Others(or myself in 6 months) may not know which pattern to follow when they make changes to this file. I'd prefer that we spend dedicate time to investigate alternative options than the "custom mock type with `XCTExpectation`" pattern used here, and update the whole test with the new pattern. What do you think?

Reply:
> I'd prefer that we spend dedicate time to investigate alternative options than the "custom mock type with XCTExpectation" pattern used here, and update the whole test [_unit tests suite?_] with the new pattern.

That would be very appropriate. It would also gives us a chance to see how the verify the different use cases.

> Others(or myself in 6 months)

So true! I don't trust my future self's memory either 😄

Reply:
Tracked in https://github.com/wordpress-mobile/WordPress-iOS/issues/19373